lib/pages/add_bridge/add_bridge.dart

```diff
@@ -805,50 +738,71 @@ class BotController extends State<AddBridge> {
     String? directChat = client.getDirectChatFromUserId(botUserId);
     directChat ??= await client.startDirectChat(botUserId);
 
-    String result = "Not logged";
-
-    while (continueProcess) {
-      result = await _checkLatestMessages(
-          directChat, successMatch, timeOutMatch, whatsAppNetwork);
-
-      if (result != "Not logged") {
-        break;
-      }
-
-      if (!continueProcess) {
-        result = "Stop Listening";
-        break;
+    final Completer<String> completer = Completer<String>();
+
+    StreamSubscription? subscription;
+    subscription = client.onEvent.stream.listen((eventUpdate) {
+      if (eventUpdate.content['sender'].contains(whatsAppNetwork.chatBot)) {
+        _onWhatsAppMessage(
+          directChat!,
+          botUserId,
+          successMatch,
+          timeOutMatch,
+          whatsAppNetwork,
+          completer,
+        );
       }
+    });
 
-      await Future.delayed(const Duration(seconds: 2));
+    try {
+      final result = await completer.future;
+      return result;
+    } finally {
+      await subscription
+          .cancel(); // Cancel the subscription to avoid memory leaks
     }
-
-    return result;
   }
 
-  Future<String> _checkLatestMessages(String directChat, RegExp successMatch,
-      RegExp timeOutMatch, SocialNetwork whatsAppNetwork) async {
-    final GetRoomEventsResponse response =
-        await client.getRoomEvents(directChat, Direction.b, limit: 3);
-    final List<MatrixEvent> latestMessages = response.chunk ?? [];
-
-    for (int i = latestMessages.length - 1; i >= 0; i--) {
-      final String messageBody =
-          latestMessages[i].content['body'].toString() ?? '';
+  /// Check last received message for WhatsApp
+  void _onWhatsAppMessage(
+      String directChat,
+      String botUserId,
+      RegExp successMatch,
+      RegExp timeOutMatch,
+      SocialNetwork whatsAppNetwork,
+      Completer<String> completer) {
+    final Room? roomBot = client.getRoomById(directChat);
+    if (roomBot == null) {
+      if (!completer.isCompleted) {
+        completer.complete("error");
```

Comment:
often times you left the error as "error". Could you please be more descriptive ? (here and in other occurences)